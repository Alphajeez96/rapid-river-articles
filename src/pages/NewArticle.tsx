import React, { useState } from 'react';
import { Link } from 'react-router-dom';
const ArticleForm: React.FC = () => {
  const [payload, setPayload] = useState({
    name: '',
    email: '',
    phone: '',
    website: '',
    title: '',
    body: '',
  });

  const [validity, setValidity] = useState({
    name: true,
    email: true,
    phone: true,
    website: true,
    title: true,
    body: true,
  });

  const handleChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    setPayload({
      ...payload,
      [e.target.name]: e.target.value,
    });

    setValidity({
      ...validity,
      [e.target.name]: true,
    });
  };

  const validateForm = () => {
    const newValidity = {
      name: payload.name.trim() !== '',
      email: /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(payload.email),
      phone: /^[0-9+-]+$/.test(payload.phone),
      website: /^(ftp|http|https):\/\/[^ "]+$/.test(payload.website),
      title: payload.title.trim() !== '',
      body: payload.body.trim() !== '',
    };

    setValidity(newValidity);
    return Object.values(newValidity).every((value) => value);
  };

  const handleForm = (e: React.FormEvent) => {
    e.preventDefault();
    const isValid = validateForm();

    if (!isValid) return;
  };
  return (
    <div className="form-holder">
      <Link to="/">
        {' '}
        <button className="h-11 w-11 rounded-full bg-primary-white flex items-center justify-center">
          <img src="/src/assets/images/caret-left.svg" alt="caret" />
        </button>
      </Link>

      <form className="card" onSubmit={handleForm}>
        <h2 className="header">Create New Article</h2>
        {/* Name - Email */}
        <div className="flex gap-8 w-full">
          <div className="form-group">
            <label htmlFor="name">Name</label>
            <input
              type="text"
              id="name"
              name="name"
              className={`input ${validity.name ? '' : 'invalid'}`}
              onChange={handleChange}
            />
          </div>

          <div className="form-group">
            <label htmlFor="email">Email</label>
            <input
              type="email"
              id="email"
              name="email"
              className={`input ${validity.email ? '' : 'invalid'}`}
              onChange={handleChange}
            />
          </div>
        </div>

        {/* Phone - Website */}
        <div className="flex gap-8 w-full">
          <div className="form-group">
            <label htmlFor="phone">Phone</label>
            <input
              type="tel"
              id="phone"
              name="phone"
              className={`input ${validity.phone ? '' : 'invalid'}`}
              onChange={handleChange}
            />
          </div>

          <div className="form-group">
            <label htmlFor="url">Website</label>
            <input
              type="url"
              id="url"
              name="website"
              className={`input ${validity.website ? '' : 'invalid'}`}
              onChange={handleChange}
            />
          </div>
        </div>

        <div className="form-group">
          <label htmlFor="title">Title</label>
          <input
            type="title"
            id="title"
            name="title"
            className={`input ${validity.title ? '' : 'invalid'}`}
            onChange={handleChange}
          />
        </div>

        <div className="form-group">
          <label htmlFor="article">Article</label>
          <textarea
            name="body"
            id="article"
            className={`input ${validity.body ? '' : 'invalid'}`}
            onChange={handleChange}
          />
        </div>

        <button className="button w-full h-14">Create</button>
      </form>
    </div>
  );
};

export default ArticleForm;
