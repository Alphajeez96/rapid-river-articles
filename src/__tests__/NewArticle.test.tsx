/* eslint-disable @typescript-eslint/no-explicit-any */
import { describe, expect, test } from 'vitest';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import NewArticle from '../pages/NewArticle';

describe('Renders the Article page', () => {
  test('snapshot', () => {
    const comp = render(
      <Router>
        <NewArticle />
      </Router>,
    );
    expect(comp).toMatchSnapshot();
  });
});
