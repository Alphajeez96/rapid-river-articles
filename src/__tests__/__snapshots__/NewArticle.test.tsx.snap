// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Renders the Article page > snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="form-holder"
      >
        <a
          href="/"
        >
           
          <button
            class="h-11 w-11 rounded-full bg-primary-white flex items-center justify-center"
          >
            <img
              alt="caret"
              src="/src/assets/images/caret-left.svg"
            />
          </button>
        </a>
        <form
          class="card"
          title="<input type="title" id="title" class="input " value="" name="title">"
        >
          <h2
            class="header"
          >
            Create New Article
          </h2>
          <div
            class="flex gap-8 w-full"
          >
            <div
              class="form-group"
            >
              <label
                for="name"
              >
                Name
              </label>
              <input
                class="input "
                id="name"
                name="name"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="email"
              >
                Email
              </label>
              <input
                class="input "
                id="email"
                name="email"
                type="email"
                value=""
              />
            </div>
          </div>
          <div
            class="flex gap-8 w-full"
          >
            <div
              class="form-group"
            >
              <label
                for="phone"
              >
                Phone
              </label>
              <input
                class="input "
                id="phone"
                name="phone"
                type="tel"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                for="url"
              >
                Website
              </label>
              <input
                class="input "
                id="url"
                name="website"
                type="url"
                value=""
              />
            </div>
          </div>
          <div
            class="form-group"
          >
            <label
              for="title"
            >
              Title
            </label>
            <input
              class="input "
              id="title"
              name="title"
              type="title"
              value=""
            />
          </div>
          <div
            class="form-group relative"
          >
            <label
              for="article"
            >
              Article
            </label>
            <textarea
              class="input "
              id="article"
              name="body"
            />
            <div
              class="absolute bottom-3 right-3"
            >
              <p
                class="text-primary-grey text-xs font-medium"
              >
                 
                0
                 /
                250
              </p>
            </div>
          </div>
          <button
            class="button w-full h-14"
          >
            Create
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="form-holder"
    >
      <a
        href="/"
      >
         
        <button
          class="h-11 w-11 rounded-full bg-primary-white flex items-center justify-center"
        >
          <img
            alt="caret"
            src="/src/assets/images/caret-left.svg"
          />
        </button>
      </a>
      <form
        class="card"
        title="<input type="title" id="title" class="input " value="" name="title">"
      >
        <h2
          class="header"
        >
          Create New Article
        </h2>
        <div
          class="flex gap-8 w-full"
        >
          <div
            class="form-group"
          >
            <label
              for="name"
            >
              Name
            </label>
            <input
              class="input "
              id="name"
              name="name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="email"
            >
              Email
            </label>
            <input
              class="input "
              id="email"
              name="email"
              type="email"
              value=""
            />
          </div>
        </div>
        <div
          class="flex gap-8 w-full"
        >
          <div
            class="form-group"
          >
            <label
              for="phone"
            >
              Phone
            </label>
            <input
              class="input "
              id="phone"
              name="phone"
              type="tel"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              for="url"
            >
              Website
            </label>
            <input
              class="input "
              id="url"
              name="website"
              type="url"
              value=""
            />
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            for="title"
          >
            Title
          </label>
          <input
            class="input "
            id="title"
            name="title"
            type="title"
            value=""
          />
        </div>
        <div
          class="form-group relative"
        >
          <label
            for="article"
          >
            Article
          </label>
          <textarea
            class="input "
            id="article"
            name="body"
          />
          <div
            class="absolute bottom-3 right-3"
          >
            <p
              class="text-primary-grey text-xs font-medium"
            >
               
              0
               /
              250
            </p>
          </div>
        </div>
        <button
          class="button w-full h-14"
        >
          Create
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
